application/src/main/java/org/togetherjava/tjbot/commands/moderation/attachment/BlacklistedAttachmentListener.java

```diff
@@ -0,0 +1,115 @@
+package org.togetherjava.tjbot.commands.moderation.attachment;
+
+import net.dv8tion.jda.api.EmbedBuilder;
+import net.dv8tion.jda.api.MessageBuilder;
+import net.dv8tion.jda.api.entities.Message;
+import net.dv8tion.jda.api.entities.MessageEmbed;
+import net.dv8tion.jda.api.events.message.MessageReceivedEvent;
+import net.dv8tion.jda.api.requests.RestAction;
+import org.jetbrains.annotations.NotNull;
+import org.togetherjava.tjbot.commands.MessageReceiverAdapter;
+import org.togetherjava.tjbot.config.Config;
+import org.togetherjava.tjbot.moderation.ModAuditLogWriter;
+
+import java.awt.*;
+import java.util.List;
+import java.util.Locale;
+import java.util.Objects;
+import java.util.regex.Pattern;
+
+/**
+ * Reacts to blacklisted attachmenents being posted, upon which they are deleted.
+ */
+public class BlacklistedAttachmentListener extends MessageReceiverAdapter {
+    private final ModAuditLogWriter modAuditLogWriter;
+    private final List<String> blacklistedFileExtensions;
+
+    /**
+     * Creates a AttachmentListener to receive all message sent in any channel.
+     *
+     * @param config to find the blacklisted media attachments
+     * @param modAuditLogWriter to inform the mods about the suspicious attachment
+     */
+    public BlacklistedAttachmentListener(@NotNull Config config,
+            @NotNull ModAuditLogWriter modAuditLogWriter) {
+        super(Pattern.compile(".*"));
+        this.modAuditLogWriter = modAuditLogWriter;
+        this.blacklistedFileExtensions = config.getBlacklistedFileExtensions();
+    }
+
+    @Override
+    public void onMessageReceived(@NotNull MessageReceivedEvent event) {
+        if (event.getAuthor().isBot() || event.isWebhookMessage()) {
+            return;
+        }
+        if (doesMessageContainBlacklistedContent(event.getMessage())) {
+            handleBadMessage(event.getMessage());
+            warnMods(event.getMessage());
```

Comment:
Needs to be one action queue, not two. U didnt fix this yet (but resolved the conversation)

Reply:
`modAuditLogWriter.write` doesnt return and RestAction it return void any idea how i can use flatMap with this or should i change the modAuditlog if say any hint how i should change it?

Reply:
u would use `map` if it doesnt return a rest action

Reply:
`   message.delete().flatMap(any -> dmUser(message)).map(any -> warnMods(message)).queue();` returns --> `no instance(s) of type variable(s) O exist so that void conforms to O`

i guess i am doing it wrong or where should i put the map on ?

Reply:
Two things:

1. why does your dmUser(message) returns a RestAction<Message>?
    if you use that in a flatMap() you end up having a RestAction<RestAction<Message>>
2. shouldn't you use the "any" in your lambdas?
3. warnMods(message) returns void but the map methods wants an actual return

Reply:
1. If i dont return and RestAction i cant chain it up in a flatmap thats why, maybe i did it wrong?
2. I got this example from the Codebase other Classes uses whit very often
3. I return void because the `modAuditLogWriter.write` returns void and no RestAction

Reply:
1. true you use flatMap it should work
2. I'm not sure
3. yea, but wasn't there some kind of method in RestAction you should use instead of map if you hava void as return?

Reply:
2. At least 5 Classes uses `flatMap(any -> method()` i am pretty sure 😄

Reply:
1. `flatMap` is used if you have a `RestAction` and want to do sth with it that results in another `RestAction`. So `firstAction.flatMap(result -> anotherAction)`. Example:
```java
user.openPrivateChannel()                           // RestAction<PrivateChannel>
  .flatMap(channel -> channel.sendMessage(message)) // RestAction<Void>
  .queue();
```
`map` is used if you want to do something that does not result in another `RestAction`. Example:
```java
channel.retrieveLastMessage()              // RestAction<Message>
  .map(message -> message.getContentRaw()) // RestAction<String>
  ...
```
2. `any` is the idiomatic way to write "i dont care what it is". for restactions, it is the result of the previous action.
```java
message1.delete()
  .flatMap(any -> message2.delete()) // dont care about the results
  .queue();                          // of the first message deletion
```

3. If u have a `RestAction` and want to do sth that returns `void`, you have 3 options:
3.1. It is the end of the queue, you do not want to do any more actions anyways, then use the consumer in `queue`:
```java
jda.retrieveUserById(123) // RestAction<User>
  .queue(user -> System.out.println(user.getName()));
```
3.2. you want to continue the chain, but you do not care about the result of the previous action, then go into `RestAction<Void>`:
```java
jda.retrieveUserById(123) // RestAction<User>
  .<Void>map(user -> {
    System.out.println(user.getName());
    return null;
  })  // RestAction<Void>
  .flatMap(any -> channel.sendMessage("Done"))
  .queue();
```
3.3. you want to continue the chain, but you do care about the result of the previous action, then pass out the input as output:
```java
jda.retrieveUserById(123) // RestAction<User>
  .<Void>map(user -> {
    System.out.println(user.getName());
    return user;
  })  // RestAction<User>
  .flatMap(any -> guild.banUser(user))
  .queue();
```